core/pkg/server/sender.go

```diff
@@ -605,13 +619,16 @@ func (s *Sender) sendRequestDefer(request *spb.DeferRequest) {
 		// Order matters: we must stop watching files first, since that pushes
 		// updates to the runfiles uploader. The uploader creates file upload
 		// tasks, so it must be flushed before we close the file transfer
-		// manager.
+		// manager. Similarly, the artifact uploader must be flushed before
+		// the file transfer manager is closed.
 		go func() {
 			defer s.logger.Reraise()
 			s.fileWatcher.Finish()
 			if s.fileTransferManager != nil {
 				s.runfilesUploader.UploadRemaining()
 				s.runfilesUploader.Finish()
+				s.artifactsSaver.Finish()
```

Comment:
is this strictly necessary since you are using a separate WG in sender.go that tracks the number of `save` calls. One reason I imagine you might want the `Finish()` is for it to disable further use of the `ArtifactSaver` in case there out of order delivery of messages?

Reply:
Good point, it's not necessary with the semantics of `Save`. Removed the method!